src/lib/mongoBackend/mongoGetSubscriptions.cpp

```diff
@@ -199,6 +201,18 @@ static void setNotification(Subscription* subP, const orion::BSONObj& r, const s
       subP->notification.timesSent += cSubP->count;
     }
 
+    if (cSubP->lastFailure != 0)
+    {
+     if (nP->counter < nP->maxLimit)
+     {
+       nP->counter = nP->counter +1;
+     }
+    }
+    else
+    {
+      nP->counter = -1;
+    }
+
```

Comment:
This is not the place where you should include counter increase. Note that mongoGetSubscriptions.cpp logic is about rendering the subscription information at DB into the JSON payload responses to GET subscriptions requests.

I think the right places should be at 

* src/lib/ngsiNotify/QueueWorkers.cpp function workerFunc()
*  src/lib/ngsiNotify/senderThread.cpp function startSenderThread()

In both case, the notification is done by httpRequestSendWithCurl() or httpRequestSend() (depending the case), getting the result value in `r` variable. You will see a structure like this in both places:

```
if (r == 0)
{
   <things we do if notification was sent correctly>
}
else
{
  <things we do if notification failed>
}
```

So:

* In the `<things we do if notification was sent correctly>` block you should include the logic that reset the counter to 0
* In the `<things we do if notification failed>` block you should include the logic that increases the counter and if the counter overpasses the limit change the status of the subscription to inactive (both in DB and in csubs cache)

Reply:
Added the counter and logic in `src/lib/ngsiNotify/QueueWorkers.cpp` and `src/lib/ngsiNotify/senderThread.cpp` but I am facing issue while comparison of failsCounter with maxFailsLimit. Could you please guide me on this? Thanks

Reply:
You should get these values through SenderThreadParams object, expanding it with the new fields you need.

Note that the SenderThreadParams object that the workerFunc() and startSenderThread() functions use to deal with the notification comes at the end from the triggered subscriptions (either in the DB or in the csubs cache) during an entity creation/update. Please have a look to the buildSenderParams() and buildSenderParamsCustom() methods, which are the ones which build the SenderThreadParams object. You should expand them to take into account the new parameters which, in sequence, will lead you to expand the TriggeredSubscription class.

You may find useful the work I'm currently doing at PR https://github.com/telefonicaid/fiware-orion/pull/3888. You can have a look to how new csub fields (mqttTopic and mqttQoS) have been added and try to do something similar for yours.

Reply:
Maybe I haven't explained correctly. I'll try again to clarify.

* Orion checks triggered subscriptions at entity creation/update. Specifically, this is done by either `addTriggeredSubscriptions_withCache()` (when csubs cache is used) or `addTriggeredSubscriptions_noCache()` (when csub cache is not used). The usage of csubs cache or not depends on the `-noCache` CLI. An important difference between both is that `_withCache()` takes the subscription for the csub cache (a data structure managed in the Orion memory) while `_noCache()` takes the subscription from the DB. Whichever the case, the corresponding subscription with all the needed information is stored as `TriggeredSubscription` object in a hashmap.
* After the add triggered subscriptions step, there is another function named `processSubscriptions()` which processed every `TriggeredSubscription` object in that hashmap. At the end, this function dispatches the notifications invoking the `sendNotifyContextRequest()` on the corresponding notifier object (`Notifier` class or `QueueNotifier` class).
* No matter which notifier, the `sendNotifyContextRequest()` builds the `SenderThreadParams` object which stores all the information that the processing thread needs to effectively sent the notification by the HTTP interface at the end. This is done by calling the `buildSenderParams()` function. Once the params have been built, then the notification is passed to the thread that will effectively sent it (either directly in the case of `Notifier` or using a notification queue in the case of `QueueNotifier`), by the `workerFunc()` or `startSenderThread()` functions, depending the case.

As you can deduct from the above, the design principle is to access to the subscription information at a single point (the add triggered subscriptions step) so the rest of the sequence is independent of that. So, you don't have to worry about if the subscription came from the DB or the csub cache when you are processing the `SenderThreadParams` to send the notification at `workerFunc()` or `startSenderThread()`. Of course, the "disable after a given number of fails functionality" has to work both in cache or not cache cases, i.e. either if Orion is started with `-noCache` or not.

I hope this time I have explained with you cannot access directly to subscriptions in DB in either `workerFunc()` or `startSenderThread()` using `SenderThreadParams` instead. A solution based on getting the subscription information again from these function would go agains our design pattern.

If you want to see a practical example of what I say, please have a look to PR #3888. In that PR I needed to "pass-through" some parameteres in the subcription (encapsulated in a `MqttInfo` object) to the final notification logic at `workerFunc()` or `startSenderThread()`. I'd suggest to have a close look to the modifications in `addTriggeredSubscriptions_withCache()`, `addTriggeredSubscriptions_noCache()`, `TriggeredSubscription`, `SenderThreadParams` and `buildSenderParams()/buildSenderParamsCustom()` to understand how this is done.

Reply:
As per your suggesstion I have added new fields in SenderThreadParams (params->maxFailsLimit and params->failsCounter) and also done modifications in `addTriggeredSubscriptions_withCache()`, `addTriggeredSubscriptions_noCache()`,` TriggeredSubscription`, `SenderThreadParams` and `buildSenderParams()/buildSenderParamsCustom()`. Also include these parametes in `workerFunc()` or `startSenderThread()` and get the values of params->maxFailsLimit and params->failsCounter.

Now I am trying to increase the params->failsCounter in else part in `src/lib/ngsiNotify/senderThread.cpp` as suggested by you in comment https://github.com/telefonicaid/fiware-orion/pull/3890#discussion_r667810559 but the failsCounter increased only once.

As per my understaing the SenderThreadParams is called only once per notification So, the failsCounter value increase only by 1 when the receiver goes down.

```
void* startSenderThread(void* p)
{
  std::vector<SenderThreadParams*>* paramsV = (std::vector<SenderThreadParams*>*) p;

  for (unsigned ix = 0; ix < paramsV->size(); ix++)
  {
    SenderThreadParams* params = (SenderThreadParams*) (*paramsV)[ix];
    char                portV[STRING_SIZE_FOR_INT];
    std::string         url;

    snprintf(portV, sizeof(portV), "%d", params->port);
    url = params->ip + ":" + portV + params->resource;

    strncpy(transactionId, params->transactionId, sizeof(transactionId));

    LM_T(LmtNotifier, ("sending to: host='%s', port=%d, verb=%s, tenant='%s',  maxFailsLimit='%lu', failsCounter='%lu', service-path: '%s', xauthToken: '%s', path='%s', content-type: %s",
                       params->ip.c_str(),
                       params->port,
                       params->verb.c_str(),
                       params->tenant.c_str(),
                       params->maxFailsLimit,
                       params->failsCounter,
                       params->servicePath.c_str(),
                       params->xauthToken.c_str(),
                       params->resource.c_str(),
                       params->content_type.c_str()));

    long long    statusCode = -1;
    std::string  out;
    LM_T(LmtNotificationRequestPayload , ("notification request payload: %s", params->content.c_str()));

    if (!simulatedNotification)
    {
      int          r;

      r = httpRequestSend(params->from,
                          params->ip,
                          params->port,
                          params->protocol,
                          params->verb,
                          params->tenant,
                          params->maxFailsLimit,
                          params->failsCounter,
                          params->servicePath,
                          params->xauthToken,
                          params->resource,
                          params->content_type,
                          params->content,
                          params->fiwareCorrelator,
                          params->renderFormat,
                          &out,
                          &statusCode,
                          params->extraHeaders);

      LM_T(LmtNotificationResponsePayload, ("notification response: %s", out.c_str()));
      if (r == 0)
      {
        __sync_fetch_and_add(&noOfNotificationsSent, 1);
        alarmMgr.notificationErrorReset(url);
        params->failsCounter = 0;

        if (params->registration == false)
        {
          subNotificationErrorStatus(params->tenant, params->subscriptionId, 0, statusCode, "");
        }
      }

      else
      {
        alarmMgr.notificationError(url, "notification failure for sender-thread: " + out);
        params->failsCounter = params->failsCounter +1;
        if (params->registration == false)
        {
          subNotificationErrorStatus(params->tenant, params->subscriptionId, -1, -1, out);
        }
      }
    }

```

 Can you please confirm my understating.

There are two methods to test the connection attempts:
1. Only send update  when subscriber is down. 
2. Sending update again and again when subscriber is down.

As per my understanding we should test according to 1.

Reply:
Thanks for the update! I have provided extra feedback in the form of new comments.

----------

With regards to 

> As per my understaing the SenderThreadParams is called only once per notification So, the failsCounter value increase only by 1 when the subscriber goes down.

I think that's correct. Each time a notification fails the increase is by 1 in the corresponding counter associated to the subscription.

------------

With regards to

> There are two methods to test the connection attempts:
> 
>     1. Only send update  when subscriber is down.
> 
>     2. Sending update again and again when subscriber is down.
> 
> 
> As per my understanding we should test according to 1.

Not sure if I understand what you mean... Anyway, the right behaviour should be the following, as far as I understand:

1. Each time the notification fails +1 is added to the fails counter associated to the subscription
2. Each time the notification goes ok the fails counter is reset to 0
3. Each time the subscription goes from "inactive" to "active" state (due to a manual intervention on the subscription API) the fail counter is reset o 0 (this is needed to "reintroduce" a given subscription once it has been disabled due to this feature)
3. If the fail counter goes beyond the max fails value, then the subscription passes to "inactive" status. That would cause no new notification gets send (nothing to implement here, that logic already exists in Context Broker).

Reply:
> Not sure if I understand what you mean... Anyway, the right behaviour should be the following, as far as I understand:
> 
> Each time the notification fails +1 is added to the fails counter associated to the subscription
> Each time the notification goes ok the fails counter is reset to 0
> Each time the subscription goes from "inactive" to "active" state (due to a manual intervention on the subscription API) the fail counter is reset o 0 (this is needed to "reintroduce" a given subscription once it has been disabled due to this feature)
> If the fail counter goes beyond the max fails value, then the subscription passes to "inactive" status. That would cause no new notification gets send (nothing to implement here, that logic already exists in Context Broker).

Fixed issue in https://github.com/telefonicaid/fiware-orion/pull/3890/commits/191b4faed1d9904e41ca7ab924c7a515934c0c4c 
Please have a look my modifications.